feat: add colcon-build and colcon-test actions (#49)
autowarefoundation/autoware-github-actions: colcon-test/README.md

# colcon-test

> This action runs `colcon test` with label.

Reviewer: ```suggestion
This action runs `colcon test` with labels.
```

Author: This action accepts only one label as input. Therefore, the plural form is not appropriate.

Reviewer: Then, let's add `a`. Also, the default value `.*` seems strange to me.

Author: ```suggestion
This action runs `colcon test` with a label.
```

Reviewer: But why can't we use multiple tags here? Are there some technical reasons?

Author: When using the CodeCov action, the label name and coverage result must be specified for each label. As a result, we will be executing as many CodeCov actions as there are labels.
In this case, we may want to use a matrix workflow. To use the matrix, I decided to use only one tag.

Reviewer: I see, thank you.

Is it impossible to use different tags for `colcon test` and `CodeCov`?
I think it can support flexible use cases.

Author: > Is it impossible to use different tags for colcon test and CodeCov?
I think it can support flexible use cases.

It is possible. If you want to execute gtest and launch_test at once, you can specify `label: "gtest|launch_test"`. And send it to CodeCov with `tags: gtest_and_launch_test` argument.

Note that you need to build target packages before running this action.

## Usage

```yaml
jobs:
  build:
    runs-on: ubuntu-latest
    container: ros:galactic
    steps:
      - name: Check out repository
        uses: actions/checkout@v2
        with:
          fetch-depth: 0

      - name: Get modified packages
        id: get-modified-packages
        uses: autowarefoundation/autoware-github-actions/get-modified-packages@tier4/proposal

>       - name: Build and test

Reviewer: ```suggestion
      - name: Build
```

        if: ${{ steps.get-modified-packages.outputs.modified-packages != '' }}
        uses: autowarefoundation/autoware-github-actions/colcon-build@tier4/proposal
        with:
          rosdistro: galactic
          target-packages: ${{ steps.get-modified-packages.outputs.modified-packages }}
          build-depends-repos: build_depends.repos

  test:

>     needs: build-and-test

Reviewer: ```suggestion
    needs: build
```

    runs-on: ubuntu-latest
    container: ros:galactic
    strategy:
      matrix:

>         label: [gtest, launch_test]

Reviewer: [IMO] This style is better since it's easier(less diff) when updating the label.

```suggestion
        label: 
          - gtest
          - launch_test
```

> 

Reviewer: [Nits] I decided not to add an empty line here.

```suggestion
```

      - name: Test
        id: test
        if: ${{ steps.get-modified-packages.outputs.modified-packages != '' }}
        uses: autowarefoundation/autoware-github-actions/colcon-test@tier4/proposal
        with:
          rosdistro: galactic
          target-packages: ${{ steps.get-modified-packages.outputs.modified-packages }}
          build-depends-repos: build_depends.repos
          label: ${{ matrix.label }}

      - name: Upload coverage to Codecov
        if: ${{ steps.test.outputs.coverage-reports != '' }}
        uses: codecov/codecov-action@v2
        with:
          files: ${{ steps.test.outputs.coverage-reports }}
          fail_ci_if_error: true
          verbose: true
          flags: ${{ matrix.label }}
```

## Inputs

| Name                | Required | Description                                     |
| ------------------- | -------- | ----------------------------------------------- |
| rosdistro           | true     | ROS distro.                                     |
| target-packages     | true     | The target packages to test.                    |
| build-depends-repos | false    | `.repos` file that includes build dependencies. |

> | label               | false    | Specify test label for test.                    |

Reviewer: [IMO]

```suggestion
| label               | false    | The label to run tests.                         |
```

# colcon-test

> This action runs `colcon test` with a label.

Reviewer: Considering this conversation, how about this?
https://github.com/autowarefoundation/autoware-github-actions/pull/49#discussion_r800471009

```suggestion
This action runs `colcon test` with labels specified in a regex format.
```

Author: Good. Thanks!

Note that you need to build target packages before running this action.

## Usage

```yaml
jobs:
  build:
    runs-on: ubuntu-latest
    container: ros:galactic
    steps:
      - name: Check out repository
        uses: actions/checkout@v2
        with:
          fetch-depth: 0

      - name: Get modified packages
        id: get-modified-packages
        uses: autowarefoundation/autoware-github-actions/get-modified-packages@tier4/proposal

      - name: Build
        if: ${{ steps.get-modified-packages.outputs.modified-packages != '' }}
        uses: autowarefoundation/autoware-github-actions/colcon-build@tier4/proposal
        with:
          rosdistro: galactic
          target-packages: ${{ steps.get-modified-packages.outputs.modified-packages }}
          build-depends-repos: build_depends.repos

  test:
    needs: build
    runs-on: ubuntu-latest
    container: ros:galactic
    strategy:

>       matrix:
>         label:
>           - gtest
>           - launch_test

Reviewer: How about preparing two variables here, for example, `test-label` and `codecov-flags`.

Probably like this?

```suggestion
      include:
        - test-label: gtest
          codecov-flags: gtest

        - test-label: launch_test
          codecov-flags: launch_test
```

Author: I think it works, I just need to fix the readme.

Author: d001cbd

      - name: Test
        id: test
        if: ${{ steps.get-modified-packages.outputs.modified-packages != '' }}
        uses: autowarefoundation/autoware-github-actions/colcon-test@tier4/proposal
        with:
          rosdistro: galactic
          target-packages: ${{ steps.get-modified-packages.outputs.modified-packages }}
          build-depends-repos: build_depends.repos
          label: ${{ matrix.label }}

      - name: Upload coverage to Codecov
        if: ${{ steps.test.outputs.coverage-reports != '' }}
        uses: codecov/codecov-action@v2
        with:
          files: ${{ steps.test.outputs.coverage-reports }}
          fail_ci_if_error: false
          verbose: true
          flags: ${{ matrix.label }}
```

## Inputs

| Name                | Required | Description                                     |
| ------------------- | -------- | ----------------------------------------------- |
| rosdistro           | true     | ROS distro.                                     |
| target-packages     | true     | The target packages to test.                    |
| build-depends-repos | false    | `.repos` file that includes build dependencies. |

> | label               | false    | The label to run tests.                         |

Reviewer: Related: https://github.com/autowarefoundation/autoware-github-actions/pull/49/files#r800473867

How about this?

```suggestion
| label               | false    | The regex pattern of test labels to be run.     |
```

Author: https://github.com/autowarefoundation/autoware-github-actions/pull/49#discussion_r800483414

Reviewer: 6dc125fdfe97c3f09e26a47616c4d12b4abf5b4a

> | label               | false    | The label to run tests.                             |

Author: https://github.com/autowarefoundation/autoware-github-actions/pull/49#discussion_r800476051

```suggestion
| label               | false    | The regex pattern of test labels to be run.         |
```

| token               | false    | The token for build dependencies.                   |

## Outputs

| Name             | Description                 |
| ---------------- | --------------------------- |

> | coverage-reports | Generated coverage reports. |

Reviewer: [IMO] I feel this is better.

```suggestion
| coverage-report-files | Generated coverage report files. |
```

Whether to add `The` or not is up to you.

Author: 015c1c1
